palace/linalg/arpack.cpp

```diff
@@ -161,88 +316,30 @@ void ArpackEigenSolver::SetWhichEigenpairs(EigenSolverBase::WhichType type)
                  "TARGET_IMAGINARY for SetWhichEigenpairs!");
       break;
   }
-}
-
-void ArpackEigenSolver::SetShiftInvert(double tr, double ti, bool precond)
-{
-  MFEM_VERIFY(!precond, "ARPACK eigenvalue solver does not support preconditioned "
-                        "spectral transformation option!");
-  sigma = tr + PETSC_i * ti;
-  sinvert = true;
-}
-
-void ArpackEigenSolver::SetInitialSpace(const petsc::PetscParVector &v)
-{
-  if (!r)
-  {
-    r = new petsc::PetscParVector(v);
-  }
-  else
-  {
-    MFEM_VERIFY(v.GetSize() == r->GetSize(),
-                "Invalid modification of eigenvalue problem size!");
-    r->Copy(v);
-  }
-  info = 1;
-}
-
-int ArpackEigenSolver::SolveInternal(petsc::PetscParVector &r_, petsc::PetscDenseMatrix &V_,
-                                     PetscScalar *eig_, int *perm_)
-{
-  MPI_Comm comm;
-  MPI_Fint fcomm;
-  a_int ido, info_ = (a_int)info;
-  a_int iparam[11] = {0}, ipntr[14] = {0};
-  a_int n, nev_, ncv_;
-  ::arpack::bmat bmat_option =
-      (opB) ? ::arpack::bmat::generalized : ::arpack::bmat::identity;
-  PetscScalar *workd, *workl;
-  double *rwork;
-  a_int lworkl;
-
-  comm = r_.GetComm();
-  fcomm = MPI_Comm_c2f(comm);
-  iparam[0] = 1;                // Exact shifts
-  iparam[2] = (a_int)max_it;    // Maximum number of Arnoldi iterations
-  iparam[3] = 1;                // Block size
-  iparam[4] = 0;                // Number of converged Ritz values
-  iparam[6] = sinvert ? 3 : 1;  // Problem mode
-
-  // Set problem sizes. The cast to int should always be safe because this is a local size.
-  n = (a_int)r_.GetSize();
-  nev_ = (a_int)nev;
-  ncv_ = (a_int)ncv;
 
   // Allocate work arrays.
-  lworkl = 3 * ncv_ * ncv_ + 5 * ncv_;
-  workd = new PetscScalar[3 * n];
-  workl = new PetscScalar[lworkl];
-  rwork = new double[ncv_];
-
-  PetscScalar *pr_ = r_.GetArray();
-  PetscScalar *pV_ = V_.GetArray();
-  petsc::PetscParVector x(comm, n, PETSC_DECIDE, nullptr);
-  petsc::PetscParVector y(comm, n, PETSC_DECIDE, nullptr);
+  a_int lworkl = 3 * ncv * ncv + 5 * ncv;
+  auto workd = std::make_unique<std::complex<double>[]>(3 * n);
+  auto workl = std::make_unique<std::complex<double>[]>(lworkl);
+  auto rwork = std::make_unique<double[]>(ncv);
```

Comment:
Totally the same thing so doesn't matter either way really. I chose a smart-pointer-managed array here rather than `std::vector` because that was the natural replacement for using `new/delete` with raw pointers before, and I didn't have need for the `std::vector` API as it's just being passed to ARPACK.

Reply:
The handling of these as buffers + sizes is on purpose to match as closely as possible the typical use of ARPACK in Fortran and other codes. It's a nice sanity check for correct usage of the library and allows people to look at our code and compare one-to-one with the names in the ARPACK API. Closing as resolved for now.